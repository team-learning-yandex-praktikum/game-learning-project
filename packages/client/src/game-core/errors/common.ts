class LogicError extends Error {
    constructor(msg: string) {
        super(`[Логическая ошибка]: ${msg}`)
    }
}

export { LogicError }
