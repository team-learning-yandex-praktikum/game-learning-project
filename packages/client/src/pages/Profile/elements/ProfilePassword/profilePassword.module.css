.title {
  margin-bottom: 80px;
  color: var(--color-accent-secondary);
  composes: font-design-headline-l from global;
}

.fields {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 40px;
}

.button {
  margin-bottom: 20px;
}
