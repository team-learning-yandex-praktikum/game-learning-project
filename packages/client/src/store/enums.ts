export enum LoadStatus {
    idle = 'idle',
    loading = 'loading',
    complete = 'complete',
    failed = 'failed',
}
