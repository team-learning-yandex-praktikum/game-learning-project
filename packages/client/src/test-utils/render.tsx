import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@store'
import { ReactNode } from 'react'

export const withProviders = (component: ReactNode) => {
    const testStore = store

    return (
        <BrowserRouter>
            <Provider store={testStore}>{component}</Provider>
        </BrowserRouter>
    )
}
