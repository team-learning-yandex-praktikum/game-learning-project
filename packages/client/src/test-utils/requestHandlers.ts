import { http, HttpResponse } from 'msw'
import { joinURL } from '@utils/url/joinURL'
import { EXTERNAL_SERVER_PATH, SERVER_URL } from '@api/baseApi/constants'
import { UserDTO } from '@api/auth/types'
import { transformData } from '@utils/data'
import { UpdatingProfileDTO } from '@api/user/types'

export const userInfoStub: UserDTO = {
    id: 111,
    login: 'ivan',
    display_name: null,
    avatar: null,
    first_name: 'Иван',
    second_name: 'Иванов',
    email: 'ivan@test.ru',
    phone: '79999999999',
}

const baseServerUrl = joinURL(SERVER_URL, EXTERNAL_SERVER_PATH)
export const requestHandlers = {
    signup: http.post(joinURL(baseServerUrl, 'auth/signup'), () =>
        HttpResponse.text(String(userInfoStub.id))
    ),
    signin: http.post(joinURL(baseServerUrl, 'auth/signin'), () =>
        HttpResponse.text('OK')
    ),
    me: http.get(joinURL(baseServerUrl, 'auth/user'), () =>
        HttpResponse.json(userInfoStub)
    ),
    logout: http.post(joinURL(baseServerUrl, 'auth/logout'), () =>
        HttpResponse.text('OK')
    ),
    updatePassword: http.put(joinURL(baseServerUrl, 'user/password'), () =>
        HttpResponse.text('OK')
    ),
    updateProfile: http.put(
        joinURL(baseServerUrl, 'user/profile'),
        async ({ request }) => {
            const body = (await request.json()) as UpdatingProfileDTO
            return HttpResponse.json(
                transformData.from.dto({
                    ...userInfoStub,
                    ...body,
                })
            )
        }
    ),
}
