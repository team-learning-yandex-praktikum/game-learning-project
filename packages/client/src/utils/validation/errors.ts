export const ERRORS = {
    required: 'Обязательное поле',
    minLength: 'Минимальное количество символов',
    maxLength: 'Максимальное количество символов',
    pattern: 'Неверный формат',
    notEquals: 'Введенные поля не совпадают',
    equalsPassword: 'Пароли совпадают',
}

export const AUTH_ERRORS = {
    authenticated: 'User already in system',
    notAuthorized: 'Request failed with status code 401',
}

export const getMinLengthError = (value: number) =>
    `${ERRORS.minLength} - ${value}`

export const getMaxLengthError = (value: number) =>
    `${ERRORS.maxLength} - ${value}`
