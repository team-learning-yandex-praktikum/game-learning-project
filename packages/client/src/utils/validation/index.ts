export { validationConfig } from './validationConfig'
export { fieldsConfig } from './fieldsConfig'
