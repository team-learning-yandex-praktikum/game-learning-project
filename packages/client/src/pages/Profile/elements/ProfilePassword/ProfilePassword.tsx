import { FC } from 'react'
import styles from './profilePassword.module.css'
import Title from '@components/Title'
import Form from '@components/Form'
import { fieldsConfig } from '@utils/validation/fieldsConfig'

export const ProfilePassword: FC = () => {
  return (
    <>
      <Title className={styles.title}>Изменить пароль</Title>
      <Form
        fields={{
          old_password: fieldsConfig.old_password,
          new_password: fieldsConfig.new_password,
          new_password_repeat: fieldsConfig.new_password_repeat,
        }}
        SubmitButtonProps={{
          children: 'Сохранить',
          variant: 'outlined',
        }}
        onSubmit={data => {
          console.log(data)
        }}
      />
    </>
  )
}

export default ProfilePassword
