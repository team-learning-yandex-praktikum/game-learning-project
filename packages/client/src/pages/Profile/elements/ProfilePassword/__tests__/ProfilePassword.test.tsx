import { setupServer } from 'msw/node'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { withProviders } from '../../../../../test-utils/render'
import { requestHandlers } from '../../../../../test-utils/requestHandlers'
import ProfilePassword from '../ProfilePassword'
import { fieldsConfig } from '@utils'
import userEvent from '@testing-library/user-event'
import { ERRORS } from '@utils/validation/errors'
import { store } from '@store'
import { LoadStatus } from '@store/enums'

const server = setupServer(requestHandlers.updatePassword)

const fieldLabels = {
    oldPassword: fieldsConfig.oldPassword.label,
    newPassword: fieldsConfig.newPassword.label,
    newPasswordRepeat: fieldsConfig.newPasswordRepeat.label,
    submit: 'Сохранить',
}

describe('Profile Password', () => {
    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())

    const testedComponent = withProviders(<ProfilePassword />)

    test('correctly rendered', () => {
        const { asFragment } = render(testedComponent)
        expect(asFragment()).toMatchSnapshot()
    })

    test('submit should be disabled', () => {
        render(testedComponent)
        const submitButton = screen.getByText(fieldLabels.submit)
        expect(submitButton).toBeDisabled()
    })

    test('submit should be enabled with correct values', async () => {
        render(testedComponent)

        const oldPassword = screen.getByLabelText(fieldLabels.oldPassword)
        const newPassword = screen.getByLabelText(fieldLabels.newPassword)
        const newPasswordRepeat = screen.getByLabelText(
            fieldLabels.newPasswordRepeat
        )
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(oldPassword, 'SomePass1234')
        await userEvent.type(newPassword, 'SomePass1111')
        await userEvent.type(newPasswordRepeat, 'SomePass1111')
        await userEvent.tab()

        expect(submitButton).toBeEnabled()
    })

    test('submit should change status', async () => {
        render(testedComponent)

        const oldPassword = screen.getByLabelText(fieldLabels.oldPassword)
        const newPassword = screen.getByLabelText(fieldLabels.newPassword)
        const newPasswordRepeat = screen.getByLabelText(
            fieldLabels.newPasswordRepeat
        )
        const submitButton = screen.getByText(fieldLabels.submit)

        expect(store.getState().user?.status).toBe(LoadStatus.idle)

        await userEvent.type(oldPassword, 'SomePass1234')
        await userEvent.type(newPassword, 'SomePass1111')
        await userEvent.type(newPasswordRepeat, 'SomePass1111')
        await userEvent.click(submitButton)

        expect(store.getState().user?.status).toBe(LoadStatus.complete)
    })

    test('submit should be disabled with same passwords', async () => {
        render(testedComponent)

        const oldPassword = screen.getByLabelText(fieldLabels.oldPassword)
        const newPassword = screen.getByLabelText(fieldLabels.newPassword)
        const newPasswordRepeat = screen.getByLabelText(
            fieldLabels.newPasswordRepeat
        )
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(oldPassword, 'SomePass1234')
        await userEvent.type(newPassword, 'SomePass1234')
        await userEvent.type(newPasswordRepeat, 'SomePass1234')
        await userEvent.tab()

        const error = await screen.findByText(ERRORS.equalsPassword)

        expect(submitButton).toBeDisabled()
        expect(error).toBeInTheDocument()
    })
})
