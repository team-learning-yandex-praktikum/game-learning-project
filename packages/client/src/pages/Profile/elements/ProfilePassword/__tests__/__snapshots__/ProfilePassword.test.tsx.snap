// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Password correctly rendered 1`] = `
<DocumentFragment>
  <h2
    class=""
  >
    Изменить пароль
  </h2>
  <div>
    <form
      id="r:id"
    >
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="oldPassword"
          >
            Старый пароль
          </label>
          <input
            id="oldPassword"
            name="oldPassword"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="newPassword"
          >
            Новый пароль
          </label>
          <input
            id="newPassword"
            name="newPassword"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="newPasswordRepeat"
          >
            Новый пароль (еще раз)
          </label>
          <input
            id="newPasswordRepeat"
            name="newPasswordRepeat"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
    </form>
    <div>
      <button
        class=""
        disabled=""
        form="r:id"
        type="submit"
      >
        Сохранить
      </button>
    </div>
  </div>
</DocumentFragment>
`;
